### PR TITLE
Take into account timeout limit when retrying queries

### PR DESCRIPTION
**Pull request by adrianna-chang-shopify**
### Motivation / Background

Building on the work done in #44576 and #44591, we extend the logic that automatically reconnects borked db connections to take into account a timeout limit. This ensures that retries + reconnects are slow-query aware, and that we don't retry queries if a given amount of time has already passed since the query was first tried.

This value will default to 5 seconds (this is the value we currently use at Shopify), but can be adjusted via the `connection_retry_timeout` config.

### Detail

This Pull Request changes the retry logic in both `#with_raw_connection` and `#reconnect!` to first check if too long has passed since the query was first attempted. If `connection_retry_timeout` time has passed, we won't retry the query.

### Additional information

Shopify is currently patching `Mysql2Adapter`'s `#execute` in order to make retries slow-query aware. These changes will help us lean fully on the upstream code and get rid of this patch.

### Checklist

Before submitting the PR make sure the following are checked:

* [X] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [X] There are no typos in commit messages and comments.
* [X] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [X] Feature branch is up-to-date with `main` (if not - rebase it).
* [X] Pull request only contains one commit for bug fixes and small features. If it's a larger feature, multiple commits are permitted but must be descriptive.
* [X] Tests are added if you fix a bug or add a feature.
* [X] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.
* [X] PR is not in a draft state.
* [x] CI is passing.

### CODE REVIEW

File: activerecord/CHANGELOG.md
Comment (eileencodes) [inline activerecord/CHANGELOG.md:4]: I feel like broken is subjective so let's just say we reconnect.

```suggestion
    reconnects database connections to take into account a timeout limit. We won't retry
```

File: activerecord/CHANGELOG.md
Comment (eileencodes) [inline activerecord/CHANGELOG.md:6]:
```suggestion
    value will default to 5 seconds, but can be changed via the `connection_retry_timeout`
```

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (eileencodes) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:174]: I just realized that Matthew's change here set the default in the adapter. I think these methods should be moved into the `HashConfig` and the `DatabaseConfig` should implement an empty method that raises NotImplementedError. I'll move the existing ones that weren't added in this PR.

**Review by eileencodes — COMMENTED** [review]
I think if you look for how we test other timeouts we can copy one of those. I'm thinking maybe with short config option and a short sleep?

Also @matthewd can you review this?

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (eileencodes) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:174]: Oh nevermind, I forgot at this point we only have the configuration hash, not the db config, so this is fine. 👍🏼

File: activerecord/test/cases/adapter_test.rb
Comment (eileencodes) [inline activerecord/test/cases/adapter_test.rb:654]: Instead of skipping, we usually wrap the tests in a conditional using `unless current_adapter?(:SQLite3Adapter)`. Also are we skipping these because the sqlite3 memory tests were failing? If it was just those that need to be skipped you can use `unless in_memory_db?` to conditionally not run these tests for that client.

File: activerecord/test/cases/adapter_test.rb
Comment (eileencodes) [inline activerecord/test/cases/adapter_test.rb:676]: Looks like postgres isn't a fan of SLEEP - maybe we can do a normal sleep without a select? Optionally we can test mysql in the mysql adapter test like this and test the postgres adapter in it's own file with postgresql specific calls.

Comment (matthewd) [reply activerecord/test/cases/adapter_test.rb:676]: Yeah I think we can probably avoid several complications at once by directly testing `with_raw_connection` -- that way we can force retryability, plus we control the block so we can just sleep + raise without needing any DB-specific handling.

(It's private, but it's "public" to the various [internal and external] adapters, so seems fair enough to test it directly.)

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:591]: We should always use monotonic time for measuring elapsed durations

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:605]: I'd rather keep both "now" calls in the same method, at the cost of slight duplication of the subtraction / greater-than.

File: activerecord/test/cases/adapter_test.rb
Comment (adrianna-chang-shopify) [reply activerecord/test/cases/adapter_test.rb:676]: Okay, perfect ❤️ Was trying to avoid going through any private methods, but this will definitely simplify things!

File: activerecord/test/cases/adapter_test.rb
Comment (adrianna-chang-shopify) [reply activerecord/test/cases/adapter_test.rb:654]: I was skipping sqlite3 because I couldn't find a good way to set up the test, given that sqlite3 doesn't have built-in SLEEP / kill mechanisms.. If we're opting for a normal sleep and testing `with_raw_connection` directly I think we should be able to test SQLite3, so let me revisit 😄

File: activerecord/CHANGELOG.md
Comment (rafaelfranca) [inline activerecord/CHANGELOG.md:1]:
```suggestion
*   Consider timeout limit when retrying a query.
```

**Review by rafaelfranca — APPROVED** [review]

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: I wonder if instead of defaulting to an arbitrary 5 seconds, we should derive the default from the read timeout ? 

e.g. `connection_retry_timeout || read_timeout * 0.95` or something.

Comment (adrianna-chang-shopify) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: That sounds like a saner default! 👍 I'll make the change 😄

Comment (adrianna-chang-shopify) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: Hm, actually I might be missing something @byroot -- where does the read timeout come from? I mistakenly thought the `timeout` from the db config was a read timeout, but I've just realized that's the timeout to check out a connection from the pool 😅

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: Right, I'm not sure if it's normalized for all adapters, I'll check.

Comment (matthewd) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: I think I kinda prefer an arbitrary 5 seconds... I'd rather not conflate it with network-behaviour expectations, where [especially as a global default] this is more of a failsafe "if we've already spent a while on it, best not retry and risk doubling that time before we return a failure".

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: Hum, so Active Record isn't actually aware of the timeout :/

For MySQL it's passed transparently to the client, as `read_timeout`:
  - https://github.com/rails/rails/blob/5fb3e721e70a13a7c0facf2d9175c15fb5132cc4/activerecord/lib/active_record/connection_adapters/mysql2_adapter.rb#L30
  - https://github.com/brianmario/mysql2#connection-options

For Postgres I actually couldn't find any reference to a timeout.

So I guess we can stay with an arbitrary default like @matthewd suggest, but IMO 5 seconds is HUGE. Most connection errors will either happen very fast or will hang forever. If it was just me I'd suggest something like `100ms`, but I suppose something more conservative like `1s` might be preferable.

Comment (adrianna-chang-shopify) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: Yeah, fair! If 1s is okay with Matthew, I'll set that to the default.

Comment (matthewd) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: We're not defining a timeout here, though: we're judging how long elapsed before something else reported a failure, and using that to decide whether to give up without retrying _an otherwise seemingly retryable_ operation. If something took a couple of seconds to get through a few stack layers and then ultimately itself timeout [having decided its other end has gone away] and report a retryable error, the whole point of this functionality is that we then want to retry it.

I see this as a last-ditch attempt to avoid wasting so much time that we timeout the web request... if your DB driver is reporting a retryable error and you can't afford to risk waiting its timeout a second time, that's the timeout you should be lowering.

One second is not a reasonable timeframe to handle even trivial misbehaviour on a remote DB connection: by the time any lower layer has decided something went wrong and recommended that we reconnect and retry, we'll have crossed our threshold and disabled retry behaviour.

I'm on the fence about whether we should have a non-nil value for this at all, tbh... I'd want something to be absolutely catastrophically wrong for us to bypass the whole retry behaviour that we've just added. Even 5s feels aggressive for that (notably, that's the value that's been in use at Shopify, where I imagine there is far more specific understanding of production network behaviours than in the average app), but it strikes me as a reasonable middle ground -- mostly because, as you note, nothing's ever actually going to be taking +/- 5 seconds: either it'll be much faster, or the lower-layer timeout will be much longer and the elapsed time we see will be in the 10+ or 60+ sec range, which is indeed "that's too broken to even consider retrying" territory.

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: > One second is not a reasonable timeframe to handle even trivial misbehaviour on a remote DB connection: by the time any lower layer has decided something went wrong and recommended that we reconnect and retry, we'll have crossed our threshold and disabled retry behaviour.

That's not really the point though. 

One of the core resiliency principle to avoid cascading failures, if to have fast failures, so it's a bad idea to retry a slow failure.

`1s` is over a hundred time more than what is necessary to get a "connection reset by peer" and similar common event when the connection became invalid while it was sitting in the pool, because the server closed the connection or similar.

Also keep in mind that for most web apps out there, the response time budget is in the 100ms range, so if you were stuck on a DB operation that took over 1s you might as well give up now to preserve the platform.

> I'm on the fence about whether we should have a non-nil value for this at all, tbh...

I'm fine with making this a optional feature if we can't agree on a default. But I think it would be good to have a reasonable default for the vast majority of users who won't know how to tune this (or that it even exists).

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: > perhaps we should avoid calling this a "timeout",

Yeah, you can call it `retry_deadline`.

Comment (matthewd) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: > for most web apps out there, the response time budget is in the 100ms range

Hrmmm, I believe most web apps have request timeout configured at 60 seconds -- 30 if they're being conservative. 

The most _popular_ apps, certainly not... but I'm far more concerned about our default config working acceptably for a near-zero-traffic app that's using a DB in a different cloud provider from their web app than I am about ensuring it's tuned to fail fast for an app with a substantial dedicated ops team. For a start, the former is far more likely to encounter recoverable connection-related errors.

If 5s has been acceptable at Shopify's scale for some time, I'm really struggling to buy that Average Joe's Web App needs a substantially tighter limit for resiliency.

Comment (matthewd) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: I guess to be more specific with an example: [in the future, after we teach more queries that they're definitionally idempotent and thus safe to retry] if we encounter an issue a couple of seconds into the DB server running a pretty-slow SQL query, I would want this code to retry. If that's too long of a query runtime for a given app, I would expect them to have configured a query timeout at a more appropriate level (which will in turn result in a non-retryable error).

Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: > I believe most web apps have request timeout configured at 60 seconds -- 30 if they're being conservative.

I'm not talking about overall request timeout, but response time budget, as in the response time you are trying to reach for the vast majority of your requests (e.g. p90).

As for request timeout, most apps use Puma, so they don't have an overall request timeout at all.

> If 5s has been acceptable at Shopify's scale for some time

It doesn't matter that much for Shopify because we have a strict semian config, so slow failure are quickly turned into fast failures. The reason for the 5s is that our monkey patch when it was introduced was quite dumb and wouldn't differentiate between query timeout and other types of errors. That 5s was basically a way to single out timeouts.

> Average Joe's Web App needs a substantially tighter limit for resiliency.

What about we name it `retry_deadline`, default it to `nil` (disabled) for now, as to unblock Adrianna, and then if any of us feel strongly about having it enabled by default we debate this in a followup PR?

**Comment by byroot** [comment]
@adrianna-chang-shopify one thing we're missing here is documentation. We need an entry in the Active Record configuration guide or something like that. Best is to look where the other global adapter options are documented and add something alongside it.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (matthewd) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:173]: > It doesn't matter that much for Shopify because we have a strict semian config, so slow failure are quickly turned into fast failures.

That's fair. (I do have my eye on teaching AR some sort of circuit-breaking behaviour as a counterpoint to these retries at some point, incidentally, but that's not particularly relevant here atm.)

> What about we name it retry_deadline, default it to nil (disabled) for now, as to unblock Adrianna, and then if any of us feel strongly about having it enabled by default we debate this in a followup PR?

Yep, that works for me; much of the reason that I was okay with 5s was the notion that it would be long enough that it would be functionally disabled in all but the most egregious of situations.

**Comment by adrianna-chang-shopify** [comment]
Thank you both for the feedback @matthewd @byroot !  I learned a lot reading through your discussion above, so I appreciate the depth of your responses there too ❤️

I've made the following changes:
- Changed `connection_retry_timeout` => `retry_deadline` and changed the verbiage around "timeouts" to be more about a pre-specified "time-elapsed deadline".
- Set the default to `nil`, so all "safely retry-able" queries will be retried by default.
- Added docs. cc @byroot I know you mentioned adding these to the AR config docs, but seems to me like these should go with the `Configuring a Database` section of the guides? 🤔 I added a section to the bottom of that portion of the docs describing both the config @matthewd added and the new deadline config. Happy to move these somewhere else if there's a more appropriate spot!

File: guides/source/configuring.md
Comment (matthewd) [inline guides/source/configuring.md:2839]:
```suggestion
an otherwise-retryable query will _not_ be retried if the specified time has elapsed while the query was first tried.
```

**Comment by matthewd** [comment]
> describing [..] the config @matthewd added

😅❤️

**Comment by byroot** [comment]
> * cc @byroot I know you mentioned adding these to the AR config docs, but seems to me like these should go with the `Configuring a Database` section of the guides?

Yeah, I just said that out of the top of my head. If after looking at both you think this is the right place, that's good for me.

On another note, the spell checker action caught some stuffs.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:595]: It would make sense to move this outside of the `begin/rescue` so that it acts as a true deadline. e.g.

If you have `retries_available = 3` and `retry_deadline = 2.0`

  - first try fail after 1s, we retry
  - second try fail after another 1s, we don't retry a third time because we're past the deadline.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:939]: Same here.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (byroot) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:595]: You can even do something like (pseudo-code):

```ruby
deadline = retry_deadline && Process.clock_gettime(Process::CLOCK_MONOTONIC) + retry_deadline
```

Then in the rescue:

```ruby
if deadline && deadline < Process.clock_gettime(Process::CLOCK_MONOTONIC) 
  # bail out
end
```

**Review by byroot — APPROVED** [review]
One relatively minor thing, after that I'm happy to merge.

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (adrianna-chang-shopify) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:595]: Oh yes , definitely doesn't make sense to reset the deadline at the start of every block 🤦‍♀️ On it!

**Review by eileencodes — APPROVED** [review]

File: activerecord/lib/active_record/connection_adapters/abstract_adapter.rb
Comment (byroot) [inline activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:610]: Doesn't matter here, but since you are likely to refactor this in your followup, I figured I'd mention it.

As a general rule, when testing multiple conditions, it's preferable to perform the faster checks first and the more expensive ones later.

In this case, `retries_available > 0` is a very simple integer comparision, while checking the deadline requires to do a syscall through `Process.clock_gettime(Process::CLOCK_MONOTONIC)`.

Again, not that it matters much in the grand scheme of things, especially here, but it's just a good general practice.

Comment (adrianna-chang-shopify) [reply activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:610]: Makes a lot of sense, thanks Jean! ❤️ I can flip the ordering when I do the refactor so we avoid the syscall if we're out of retries anyway.

**Pull request merged**
